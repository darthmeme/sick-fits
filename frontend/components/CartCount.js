import React from 'react'
import styled from 'styled-components'
import { TransitionGroup, CSSTransition } from 'react-transition-group'

const Dot = styled.div`
  background: ${props => props.theme.red};
  color: white;
  line-height: 2rem;
  min-width: 3rem;
  border-radius: 50%;
  padding: 0.5rem;
  margin-left: 1rem;
  font-weight: 400;
  font-feature-settings: "tnum";
  font-variant-numeric: tabular-nums;
`

const AnimationStyles = styled.span`
  position: relative;
  
  .count {
    display: block;
    position: relative;
    backface-visibility: hidden;
    transition: all 0.3s ease;
  }

  .count-enter {
    transform: rotateX(0.5turn)
  }

  .count-enter-active {
    transform: rotateX(0)
  }

  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }

  .count-exit-active {
    transform: rotateX(0.5turn);
  }
`

const CartCount = ({ count }) => (
  <AnimationStyles>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        className="count"
        classNames="count"
        key={count}
        timeout={{ enter: 300, exit: 300 }}>
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyles>
)

export default CartCount
