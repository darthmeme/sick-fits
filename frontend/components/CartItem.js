import React, { Component } from 'react'
import formatMoney from '../lib/formatMoney'
import styled from 'styled-components'
import RemoveFromCart from './RemoveFromCart'

const CartItemStyled = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightGrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;

  img {
    width: 100px;
    margin-right: 10px;
  }

  h3, p {
    margin: 0;
  }
`

const CartItem = ({ item }) => (
  <CartItemStyled>
    <img src={item.item.image} />
    <div>
      <h3>{item.item.name}</h3>
      <p>
        {formatMoney(item.item.price * item.quantity)}
        {' - '}
        <em>
          {item.quantity} &times; {formatMoney(item.item.price)} each
        </em>
      </p>
    </div>
    <RemoveFromCart id={item.id} />
  </CartItemStyled>
)

export default CartItem
