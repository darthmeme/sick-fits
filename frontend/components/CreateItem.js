import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import gql from 'graphql-tag'
import ErrorMessage from './ErrorMessage'
import Router from 'next/router'

export const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $name: String!,
    $description: String!,
    $price: Int!,
    $image: String,
    $largeImage: String
  ) {
    createItem(
      name: $name
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
      name
    }
  }
`

export default class CreateItem extends Component {
  state = {
    name: 'Cool',
    description: '',
    image: '',
    largeImage: '',
    price: 0
  }

  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value

    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sickfits')

    const res = await fetch('https://api.cloudinary.com/v1_1/dug281bvy/image/upload', {
      method: 'POST',
      body: data
    })

    const file = await res.json()

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url
    })
  }

  render () {
    return (
      <Mutation
        mutation={CREATE_ITEM_MUTATION}
        variables={this.state}>
        {(createItem, { data, error, loading }) => (
          <Form onSubmit={
            async e => {
              e.preventDefault()
              
              const res = await createItem()
              Router.push({ pathname: '/item', query: { item: res.data.createItem.id } })
            }
          }>
            <ErrorMessage error={error} />
            <h2>Sell an item</h2>
            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="name">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  onChange={this.uploadFile} />

                {this.state.image && <img src={this.state.image} />}
              </label>

              <label htmlFor="name">
                Name
                <input
                  type="text"
                  id="name"
                  name="name"
                  placeholder="name"
                  value={this.state.name}
                  onChange={this.handleChange}
                  required />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  type="text"
                  id="description"
                  name="description"
                  placeholder="description"
                  value={this.state.description}
                  onChange={this.handleChange}
                  required />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="price"
                  value={this.state.price}
                  onChange={this.handleChange}
                  required />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}
