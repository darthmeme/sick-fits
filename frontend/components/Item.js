import React, { Component } from 'react'
import PropTypes from 'prop-types'
import Title from './styles/Title'
import ItemStyles from './styles/ItemStyles'
import PriceTag from './styles/PriceTag'
import Link from 'next/link'
import formatMoney from '../lib/formatMoney'
import DeleteItem from './DeleteItem'

export default class Item extends Component {
  static propTypes = {
    item: PropTypes.object.isRequired
  }
  
  render () {
    const { item } = this.props

    return (
      <ItemStyles>
        {item.image && <img src={item.image} />}
        <Title>
          <Link href={{ pathname: '/item', query: { id: item.id } }}>
            <a>{item.name}</a>
          </Link>
        </Title>
        <p>{item.description}</p>
        <div>
          <Link href={{ pathname: '/update', query: { id: item.id } }}><a>Edit ✏️</a></Link>
          <button>Add to card</button>
          <DeleteItem id={item.id} />
        </div>
        <PriceTag>{formatMoney(item.price)}</PriceTag>
      </ItemStyles>
    )
  }
}
