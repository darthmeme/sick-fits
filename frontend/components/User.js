import { Query } from 'react-apollo'
import gql from 'graphql-tag'

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER_QUERY {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          name
          price
          description
          image
        }
      }
    }
  }
`

const User = props => (
  <Query {...props} query={CURRENT_USER_QUERY}>
    {payload => props.children(payload)}
  </Query>
)

export default User
