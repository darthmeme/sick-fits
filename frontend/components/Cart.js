import React from 'react'
import CartStyles from './styles/CartStyles'
import CloseButton from './styles/CloseButton'
import Supreme from './styles/Supreme'
import SickButton from './styles/SickButton'
import { Query, Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import User from './User'
import CartItem from './CartItem'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'
import { adopt } from 'react-adopt'
import Payment from './Payment'

export const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`

const Composed = adopt({
  user: <User />,
  toggleCart: <Mutation mutation={TOGGLE_CART_MUTATION} />,
  localState:  <Query query={LOCAL_STATE_QUERY} />
})

const Cart = () => {
  return (
    <Composed>
      {({ user, toggleCart, localState }) => {
        const me = user.data.me

        if (!me) return null

        return (
          <CartStyles open={localState.data.cartOpen}>
            <header>
              <CloseButton
                title='Close'
                onClick={toggleCart}>
                &times;
              </CloseButton>
              <Supreme>{me.name}'s Cart</Supreme>
              <p>You have {me.cart.length} item{me.cart.length === 1 ? '' : 's'} in your cart</p>
            </header>
            <ul>
              {me.cart && me.cart.map(item => (
                <CartItem item={item} key={item.id} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(me.cart))}</p>
              {me.cart.length && (
                <Payment>
                  <SickButton>Checkout</SickButton>
                </Payment>
              )}
            </footer>
          </CartStyles>
        )
      }}
    </Composed>
  )
}

export default Cart
