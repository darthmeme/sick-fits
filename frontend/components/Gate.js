import { Query } from 'react-apollo'
import { CURRENT_USER_QUERY } from './User'
import Signin from './Signin'

const Gate = props => (
  <Query query={CURRENT_USER_QUERY}>
    {(({ data, loading }) => {
      if (loading) return <div>Loading...</div>
      if (!data.me) return <Signin />

      return props.children
    })}
  </Query>
)

export default Gate
