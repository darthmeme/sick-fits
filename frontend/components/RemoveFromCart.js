import React, { Component } from 'react'
import styled from 'styled-components'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import { CURRENT_USER_QUERY } from './User'

const REMOVE_FROM_CART_MUTATION = gql`
  mutation REMOVE_FROM_CART_MUTATION($id: ID!) {
    removeFromCart(id: $id) {
      id
    }
  }
`

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: none;
  cursor: pointer;

  &:hover {
    color: ${props => props.theme.red}
  }
`

export default class RemoveFromCart extends Component {
  update = (cache, payload) => {
    const data = cache.readQuery({ query: CURRENT_USER_QUERY })
    const id = payload.data.removeFromCart.id

    const newData = {
      me: {
        ...data.me,
        cart: data.me.cart.filter(item => item.id !== id)
      }
    }

    cache.writeQuery({ query: CURRENT_USER_QUERY, data: newData })
  }

  render () {
    return (
      <Mutation
        mutation={REMOVE_FROM_CART_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
        optimisticResponse={
          {
            __typename: 'Mutation',
            removeFromCart: {
              __typename: 'CartItem',
              id: this.props.id
            }
          }
        }>
        {(removeFromCart, { error, loading }) => (
          <BigButton
            disabled={loading}
            onClick={() => {
              removeFromCart()
            }}>
            &times;
          </BigButton>
        )}
      </Mutation>
    )
  }
}
