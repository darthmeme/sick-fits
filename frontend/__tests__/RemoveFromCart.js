import { mount } from 'enzyme'
import wait from 'waait'
import toJSON from 'enzyme-to-json'
import { MockedProvider } from 'react-apollo/test-utils'
import RemoveFromCart, { REMOVE_FROM_CART_MUTATION } from '../components/RemoveFromCart'
import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser, fakeCartItem } from '../lib/testUtils'
import { ApolloConsumer } from 'react-apollo';

const mocks = [
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem({ id: '123' })]
        }
      }
    }
  },
  {
    request: {
      query: REMOVE_FROM_CART_MUTATION,
      variables: {
        id: '123'
      }
    },
    result: {
      data: {
        removeFromCart: {
          __typename: 'CartItem',
          id: '123'
        }
      }
    }
  }
]

describe('<RemoveFromCart />', () => {
  it('renders and matches snapshot', () => {
    const wrapper = mount(
      <MockedProvider>
        <RemoveFromCart />
      </MockedProvider>
    )

    expect(toJSON(wrapper.find('button'))).toMatchSnapshot()
  })

  it('removes item from cart', async () => {
    let apolloClient
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client

            return <RemoveFromCart id='123' />
          }}
        </ApolloConsumer>
      </MockedProvider>
    )

    const res = await apolloClient.query({ query: CURRENT_USER_QUERY })
    expect(res.data.me.cart).toHaveLength(1)

    wrapper.find('button').simulate('click')

    await wait()

    const newRes = await apolloClient.query({ query: CURRENT_USER_QUERY })
    
    expect(newRes.data.me.cart).toHaveLength(0)
  })
})
