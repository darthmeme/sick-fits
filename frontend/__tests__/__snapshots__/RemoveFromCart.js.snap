// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RemoveFromCart /> renders and matches snapshot 1`] = `
<button
  className="RemoveFromCart__BigButton-emvtd6-0 dXjKBJ"
  disabled={false}
  onClick={[Function]}
>
  Ã—
</button>
`;
